pkg/downsample/downsample.go

```diff
@@ -0,0 +1,168 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package downsample
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/timescale/promscale/pkg/internal/day"
+	"github.com/timescale/promscale/pkg/log"
+	"github.com/timescale/promscale/pkg/util"
+)
+
+const (
+	setDownsamplingStateSQL       = "SELECT prom_api.set_downsampling_state($1)"
+	createOrUpdateDownsamplingSQL = "CALL _prom_catalog.create_or_update_downsampling($1, $2, $3)"
+	updateDownsamplingStateForSQL = "SELECT _prom_catalog.update_downsampling_state($1, $2)"
+	downsamplePrefix              = "ds_"          // Stands of downsample_
+	lockID                        = 55851985173278 // Choosen randomly
+)
+
+type Config struct {
+	Interval      day.Duration `yaml:"interval"`
+	Retention     day.Duration `yaml:"retention"`
+	shouldRefresh bool
+}
+
+func (c Config) Name() string {
+	return downsamplePrefix + c.Interval.Text()
+}
+
+func SetState(ctx context.Context, conn *pgx.Conn, state bool) error {
+	_, err := conn.Exec(ctx, setDownsamplingStateSQL, state)
+	if err != nil {
+		return fmt.Errorf("error setting downsampling state: %w", err)
+	}
+	return nil
+}
+
+// Sync updates the downsampling cfgs in the DB in accordance with the given new cfgs. It:
+//  1. Creates of new downsampling cfgs that are not in the database
+//  2. Updates retention duration of downsampling cfgs that are present in the database but with a different retention duration
+//  3. Enables refreshing of downsampling cfgs in the database that are in the new cfgs but were previously disabled
+//  4. Disables refreshing of downsampling cfgs in the database that were not found in the new cfgs
+func Sync(ctx context.Context, conn *pgx.Conn, cfgs []Config) error {
+	pgLock, err := util.NewPgAdvisoryLock(lockID, conn.Config().ConnString())
+	if err != nil {
+		return fmt.Errorf("error getting lock for syncing downsampling config")
+	}
+	defer pgLock.Close()
+	got, err := pgLock.GetAdvisoryLock() // To prevent failure when multiple Promscale start at the same time.
+	if err != nil {
+		return fmt.Errorf("error trying pg advisory_lock")
+	}
+	if !got {
```

Comment:
I believe this lock is used for other purposes as well. So I'd recommend waiting/blocking until lock is taken an configuration is applied. We can obviously add some timeout and error if we were unable to take lock and apply downsampling configuration.